internal/impl/gcp/input_bigquery_select.go

```diff
@@ -0,0 +1,293 @@
+			`
+Here we query the public corpus of Shakespeare's works to generate a stream of the top 10 words that are 3 or more characters long:`,
+			`
+input:
+  gcp_bigquery_select:
+    project: gc-lab-growth-0d4d
+    table: bigquery-public-data.samples.shakespeare
+    columns:
+      - word
+      - sum(word_count) as total_count
+    where: length(word) >= ?
+    suffix: |
+      GROUP BY word
+      ORDER BY total_count DESC
+      LIMIT 10
+    args_mapping: |
+      root = [ 3 ]
+`,
+		)
+}
+
+type bigqueryIterator interface {
+	Next(dst interface{}) error
+}
+
+type bigQuerySelectInput struct {
+	logger *service.Logger
+	config *bigQuerySelectInputConfig
+
+	// Used for overriding client options in tests
+	clientOpts []option.ClientOption
+	client     *bigquery.Client
+
+	shutdownSig *shutdown.Signaller
+
+	// Represents a row iterator that returns query results
+	// The indirection provided by the `bigqueryIterator` interface allows test
+	// code to conveniently create mock iterators
+	iterator bigqueryIterator
+}
+
+func newBigQuerySelectInput(inConf *service.ParsedConfig, logger *service.Logger) (*bigQuerySelectInput, error) {
+	conf, err := bigQuerySelectInputConfigFromParsed(inConf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	return &bigQuerySelectInput{
+		logger:      logger,
+		config:      &conf,
+		shutdownSig: shutdown.NewSignaller(),
+	}, nil
+}
+
+func (inp *bigQuerySelectInput) Connect(ctx context.Context) error {
+	client, err := bigquery.NewClient(context.Background(), inp.config.project, inp.clientOpts...)
+	if err != nil {
+		return fmt.Errorf("failed to create bigquery client: %w", err)
+	}
+
+	inp.client = client
+
+	query, err := inp.buildQuery()
+	if err != nil {
+		return fmt.Errorf("failed to build query: %w", err)
+	}
+
+	inp.logger.Debugf("executing query: %s", query.Q)
+
+	jobctx, _ := inp.shutdownSig.CloseAtLeisureCtx(context.Background())
+
+	job, err := query.Run(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %w", err)
+	}
+
+	logger := inp.logger.With("job_id", job.ID())
+
+	logger.Debug("running bigquery job")
+
+	status, err := job.Wait(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to wait on job: %w", err)
+	}
+
+	if err := status.Err(); err != nil {
+		return fmt.Errorf("failed to complete job: %w", err)
+	}
+
+	logger.Debug("bigquery job completed")
+
+	it, err := job.Read(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to read rows: %w", err)
+	}
+
+	inp.iterator = it
+
+	return nil
+}
+
+func (inp *bigQuerySelectInput) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	if inp.iterator == nil {
+		return nil, nil, fmt.Errorf("query result iterator is not set: %w", service.ErrNotConnected)
+	}
+
+	var row map[string]bigquery.Value
+	err := inp.iterator.Next(&row)
+	if err == iterator.Done {
+		return nil, nil, service.ErrEndOfInput
+	}
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
**Please note:** Based on conversations with @Jeffail, I am assuming `Read` is not called by multiple goroutines and therefore accessing the `*RowIterator` without syncronisation is safe. The `*RowIterator` type does not appear safe for concurrent access from my cursory read of the GCP Go SDK. If I got this wrong, then I can submit another commit to introduce a mutex around this block of code.